keccak256/src/gates/base_conversion.rs

```diff
@@ -0,0 +1,317 @@
+        &self,
+        layouter: &mut impl Layouter<F>,
+        input: (Cell, F),
+        parent_flag: (Cell, F),
+    ) -> Result<(Cell, F), Error> {
+        let (input_coefs, output_coefs, _) = self.bi.compute_coefs(input.1)?;
+
+        let (out_cell, out_value) = layouter.assign_region(
+            || "Base conversion",
+            |mut region| {
+                let mut input_acc = F::zero();
+                let input_pob = self.bi.input_pob();
+                let mut output_acc = F::zero();
+                let output_pob = self.bi.output_pob();
+                for (offset, (&input_coef, &output_coef)) in
+                    input_coefs.iter().zip(output_coefs.iter()).enumerate()
+                {
+                    self.q_lookup.enable(&mut region, offset)?;
+                    if offset != 0 {
+                        self.q_running_sum.enable(&mut region, offset)?;
+                    }
+                    let flag_cell = region.assign_advice(
+                        || "flag",
+                        self.flag,
+                        offset,
+                        || Ok(parent_flag.1),
+                    )?;
+                    region.constrain_equal(flag_cell, parent_flag.0)?;
+                    let input_coef_cell = region.assign_advice(
+                        || "Input Coef",
+                        self.input_coef,
+                        offset,
+                        || Ok(input_coef),
+                    )?;
+                    input_acc = input_acc * input_pob + input_coef;
+                    let input_acc_cell = region.assign_advice(
+                        || "Input Acc",
+                        self.input_acc,
+                        offset,
+                        || Ok(input_acc),
+                    )?;
+                    let output_coef_cell = region.assign_advice(
+                        || "Output Coef",
+                        self.output_coef,
+                        offset,
+                        || Ok(output_coef),
+                    )?;
+                    output_acc = output_acc * output_pob + output_coef;
+                    let output_acc_cell = region.assign_advice(
+                        || "Output Acc",
+                        self.output_acc,
+                        offset,
+                        || Ok(output_acc),
+                    )?;
+
+                    if offset == 0 {
+                        // bind first acc to first coef
+                        region
+                            .constrain_equal(input_acc_cell, input_coef_cell)?;
+                        region.constrain_equal(
+                            output_acc_cell,
+                            output_coef_cell,
+                        )?;
+                    } else if offset == input_coefs.len() - 1 {
+                        region.constrain_equal(input_acc_cell, input.0)?;
+                        return Ok((output_acc_cell, output_acc));
+                    }
```

Comment:
There's a length missmatch here. 2 constraints in one branch of the `if` and 1 in the `else`. 

Could you confirm if this is an issue @therealyingtong ??

Reply:
This branches on the offset. Just want a difference on offset 0 and the last offset. 

I think what might matter is how can we be sure the prover and the verifier can drive the same number of rows. The answer is they should have exactly `bi.slice_count()` rows. `input_coefs.len() == output_coefs.len() == bi.slice_count()`. But not sure how to express that here is better.

Reply:
Branching on the offset looks good to me, since we can deterministically compute that.